File: Basic/src/main/java/Generic1.java
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;

/**
 * Created by tony on 2019-06-17.
 */
public class Generic1 {

    static class InnerClass<T> {

    }

    public static void main(String[] args) {

        InnerClass<Integer> innerClass = new InnerClass<Integer>(){

        }; // 匿名内部类的声明在编译时进行，实例化在运行时进行

        Type typeClass = innerClass.getClass().getGenericSuperclass();
        System.out.println(typeClass);

        if (typeClass instanceof ParameterizedType) {
            Type actualType = ((ParameterizedType)typeClass).getActualTypeArguments()[0];

            System.out.println(actualType);
        }
    }
}

File: Basic/src/main/kotlin/generic/Generic1.kt
package generic

import java.lang.reflect.ParameterizedType

/**
 * Created by tony on 2019-06-17.
 */

object Generic1 {

    open class InnerClass<T>

    @JvmStatic
    fun main(args: Array<String>) {

        val innerClass = object : InnerClass<Int>() {

        } // 匿名内部类的声明在编译时进行，实例化在运行时进行

        val typeClass = innerClass.javaClass.genericSuperclass
        println(typeClass)

        if (typeClass is ParameterizedType) {

            val actualType = typeClass.actualTypeArguments[0]
            println(actualType)
        }
    }
}